src/rtos/FreeRTOS.c

```diff
@@ -1114,8 +1113,12 @@ static int freertos_get_current_thread_registers(struct rtos *rtos, int64_t thre
 		return ERROR_FAIL;
 	}
 
+	bool is_xtensa = strncmp(target_get_gdb_arch(current_target), "xtensa", 6) == 0;
 	for (int i = 0; i < *num_regs; i++) {
-		(*reg_list)[i].number = gdb_reg_list[i]->number;
```

Comment:
@ianstcdns  can we somehow fill `number` variable  in order to compatible to riscv or arm archs.  If I am not wrong, `target_get_gdb_reg_list` behaves different for `xtensa`

Reply:
I probably removed this for sparse register mode (for generic xtensa) before we added support for contiguous register mode (esp chips).  Fine to add this number back in, but it may only be useable for contiguous mode--could you add a check for that or at least a comment indicating that more changes may be required for sparse mode?

Reply:
`reg_list[didx].number = didx;` can work if I change the order of the registers in `xtensa-core-esp32xx.cfg` I can put the xtensa_regs first then the others. I will try and let you know.

Reply:
Unfortunately not worked. gdb reads the regs  in the same order with e.g. https://github.com/espressif/xtensa-overlays/blob/master/xtensa_esp32s3/gdb/gdb/xtensa-config.c 

I guess, you are talking about `gpio_out`.  These numbers are ok in our tests for both esp32s2 and esp32s3 
https://github.com/espressif/xtensa-overlays/blob/master/xtensa_esp32s3/gdb/gdb/xtensa-config.c#L208
https://github.com/espressif/xtensa-overlays/blob/master/xtensa_esp32s2/gdb/gdb/xtensa-config.c#L125

Reply:
What about adding a new field to the `struct xtensa_reg_desc` called `gpkt_index`  and extend `xtensa xtreg` command with the index. 
```
xtensa xtreg    pc      0x0020      0 
...
xtensa xtreg	ar63    0x013f      64
xtensa xtreg	lbeg    0x0200      65
...
```

We can add `xtensa_regs` gpkt_indexes statically in the struct init.

Then during command parse inside `COMMAND_HELPER(xtensa_cmd_xtreg_do, struct xtensa *xtensa)`
```c
if (is_extended_reg) {
    rptr->gpkt_index = gpkt_index;
}
```

Then inside `xtensa_build_reg_cache` 
```c
reg_list[didx].number = rlist[i].gpkt_index;
```

How it sounds?

Reply:
Or instead of adding a new command parameter, we can restrict the order of `xtensa xtreg` commands and fill `gpt_index` manually during command parse.

Reply:
Sorry I didn't respond yesterday.  There is a way to do this (at least there should be): when "xtregfmt contiguous" is specified (must be before the first xtreg definition), you can pass it the parameter [numgregs] which is the number of registers to include in a G-packet reply.  The first _numgregs_ registers defined in the config file will be included, and their order should match what is in the .cfg file.  See xtensa->contiguous_regs_desc at the bottom of xtensa_cmd_xtreg_do().  

If the G-register response packet is in the correct order but not the rtos register spec, would it be reasonable to reorder the register cache such that the first _numgregs_ matches the contiguous_regs_desc ordering?

Reply:
As I see, now register cache is ordering xtensa_regs + xtensa_opt_regs, but I want it  to re-order by the same as cfg file.  Then we can easily use `didx` index for the `number`

Reply:
Yes, that could work, as long as there aren't too many places in the target layer that expect the xtensa_regs + xtensa_opt_regs ordering.  I think the fetch and flush routines may have some assumptions that the ordering of the cache includes all hard-coded registers before all optional registers.  It might be easiest if you can keep this, and just rearrange the ordering within the two lists.  Or is that insufficient?  If so, the logic to fetch and flush may need to change to match register names/offsets/types instead of just index.

Reply:
Ok this one line addition works for us.  Look at the `reg_cache->reg_list[j].number = i;  // >>> This is OK. `

```c
	/* Construct contiguous register list from contiguous descriptor list */
	if (xtensa->regmap_contiguous && xtensa->contiguous_regs_desc) {
		xtensa->contiguous_regs_list = calloc(xtensa->total_regs_num, sizeof(struct reg *));
		if (!xtensa->contiguous_regs_list) {
			LOG_TARGET_ERROR(target, "ERROR: Out of memory");
			goto fail;
		}
		for (unsigned int i = 0; i < xtensa->total_regs_num; i++) {
			unsigned int j;
			for (j = 0; j < reg_cache->num_regs; j++) {
				if (!strcmp(reg_cache->reg_list[j].name, xtensa->contiguous_regs_desc[i]->name)) {
					reg_cache->reg_list[j].number = i;  // >>> This is OK.
					xtensa->contiguous_regs_list[i] = &(reg_cache->reg_list[j]);
					LOG_TARGET_DEBUG(target,
						"POPULATE contiguous regs list: %-16s, dbreg_num 0x%04x",
						xtensa->contiguous_regs_list[i]->name,
						xtensa->contiguous_regs_desc[i]->dbreg_num);
					break;
				}
			}
			if (j == reg_cache->num_regs)
				LOG_TARGET_WARNING(target, "contiguous register %s not found",
					xtensa->contiguous_regs_desc[i]->name);
		}
	}
```

Reply:
oh cool!  thanks for finding this...nice not to have to change target layer logic if not necessary.

Reply:
Should I create a patch? 
Also, maybe we can consider the rename `contiguous_regs_desc` with sth like `total_regs_desc` 
Currently it stores all the registers.

Reply:
I will try to test this change out later this afternoon; up to you on when you'd like to submit the patch.  For renaming, it's a little tricky, since the "contiguous" part of the name is intentionally there to imply a relationship to xtensa->regmap_contiguous, contiguous_regs_list, etc.  What about "contiguous_total_regs_desc" instead, if that's not too long?

Reply:
`contiguous_total_regs_desc` also leaves me an impression about keeping only given number of regs. e.g. 105
```
xtensa xtregfmt	contiguous			105
```
OK. It was just a suggestion. We can leave it as is.